src/scalar.rs

```diff
@@ -2,101 +2,90 @@
 //!
 //! **Note**: pointers are never scalars (like SPIR-V, but unlike other IRs).
 
-// HACK(eddyb) this could be some `struct` with private fields, but this `enum`
-// is only 2 bytes in size, and has better ergonomics overall.
+#[derive(Copy, Clone, PartialEq, Eq, Hash)]
+pub enum TypeKind {
+    Bool,
+    Float,
+    UInt,
+    SInt,
+}
+
 #[derive(Copy, Clone, PartialEq, Eq, Hash)]
 pub enum Type {
     Bool,
-    SInt(IntWidth),
-    UInt(IntWidth),
-    Float(FloatWidth),
+    S8,
+    S16,
+    S32,
+    S64,
+    S128,
+    U8,
+    U16,
+    U32,
+    U64,
+    U128,
+    F16,
+    F32,
+    F64,
 }
```

Comment:
This is interesting in a way I hadn't considered.
I might be open to this if the actual hardcoded list was a private `enum` and `Type` was a `struct` newtype around it, what do you think?

For the record, the reason `Type` being 2 bytes is slightly suboptimal for me, is that you can actually pack `(TypeKind, IntWidth)` into 5 bits, and use the remaining 3 bits to also support vector types or something. It's not that important, just a thought exercise I started ages ago (for LLVM IR types, once you remove pointee types, and arrays/structs, because the only nesting left is vector types around scalar types).

Reply:
I might be open to this if the actual hardcoded list was a private enum and Type was a struct newtype around it, what do you think?

Did that in https://github.com/EmbarkStudios/spirt/commit/1bdf588e2da2f3384f8b2f0f24b421ea0c412bd4 - what do you think?